Add basic validator.

marcenacp opened the pull request:
This is the first version. For the moment, I prototyped using
`datasets_format/datasets/titanic/metadata.json`.

Design elements:

- The entry is the script `datasets_format/python/format/scripts/validate.py`.
- README in `datasets_format/python/format/README.md`.
- `Dataset` represents a graph from a file.
- In `Dataset.__init__`, we call the validator.
- The validator builds an RDFLib graph, then converts it to a NetworkX
  graph, then checks this graph.
- At the moment, we only check that each subject has the proper
  properties.
- I put in place some unit tests, particularly in `datasets_format/python/format/src/graphs_test.py`.

Questions:

- What should I expect on `http://mlcommons.org/schema/dataType`? Just that the dataType is a valid URI? Should we fetch the URI?

Todo:

- Check the ID, and run a BFS to check that the tree is valid.
- Launch tests in CI/CD.
- Support more complex use cases (e.g., `ml:RecordSet` -> `data` in Titanic), and other datasets.
- Make OpenML converter compatible (for instance, at the moment, I have the following errors):

```bash
format.src.errors.ValidationError: Found the following 6 error(s) during the validation:
  -  [dataset(Titanic)] Property "https://schema.org/citation" is mandatory, but does not exist. Existing properties: {rdflib.term.URIRef('https://schema.org/version'), rdflib.term.URIRef('https://schema.org/keywords'), rdflib.term.URIRef('https://schema.org/distribution'), rdflib.term.URIRef('https://schema.org/dateModified'), rdflib.term.URIRef('https://schema.org/description'), rdflib.term.URIRef('https://schema.org/url'), rdflib.term.URIRef('https://schema.org/creativeWorkStatus'), rdflib.term.URIRef('http://www.w3.org/1999/02/22-rdf-syntax-ns#type'), rdflib.term.URIRef('https://schema.org/@language'), rdflib.term.URIRef('https://schema.org/recordSet'), rdflib.term.URIRef('https://schema.org/isAccessibleForFree'), rdflib.term.URIRef('https://schema.org/name'), rdflib.term.URIRef('https://schema.org/dateCreated')}.
  -  [dataset(Titanic)] Property "https://schema.org/license" is mandatory, but does not exist. Existing properties: {rdflib.term.URIRef('https://schema.org/version'), rdflib.term.URIRef('https://schema.org/keywords'), rdflib.term.URIRef('https://schema.org/distribution'), rdflib.term.URIRef('https://schema.org/dateModified'), rdflib.term.URIRef('https://schema.org/description'), rdflib.term.URIRef('https://schema.org/url'), rdflib.term.URIRef('https://schema.org/creativeWorkStatus'), rdflib.term.URIRef('http://www.w3.org/1999/02/22-rdf-syntax-ns#type'), rdflib.term.URIRef('https://schema.org/@language'), rdflib.term.URIRef('https://schema.org/recordSet'), rdflib.term.URIRef('https://schema.org/isAccessibleForFree'), rdflib.term.URIRef('https://schema.org/name'), rdflib.term.URIRef('https://schema.org/dateCreated')}.
  -  [dataset(Titanic) > distribution(Titanic (arff))] Property "https://schema.org/contentSize" is mandatory, but does not exist. Existing properties: {rdflib.term.URIRef('https://schema.org/name'), rdflib.term.URIRef('https://schema.org/encodingFormat'), rdflib.term.URIRef('https://schema.org/contentUrl'), rdflib.term.URIRef('http://www.w3.org/1999/02/22-rdf-syntax-ns#type')}.
  -  [dataset(Titanic) > distribution(Titanic (arff))] At least one of these properties should be defined: [rdflib.term.URIRef('https://schema.org/sha256'), rdflib.term.URIRef('https://schema.org/md5')]. Existing properties: {rdflib.term.URIRef('https://schema.org/name'), rdflib.term.URIRef('https://schema.org/encodingFormat'), rdflib.term.URIRef('https://schema.org/contentUrl'), rdflib.term.URIRef('http://www.w3.org/1999/02/22-rdf-syntax-ns#type')}.
  -  [dataset(Titanic) > distribution(Titanic (parquet))] Property "https://schema.org/contentSize" is mandatory, but does not exist. Existing properties: {rdflib.term.URIRef('https://schema.org/name'), rdflib.term.URIRef('https://schema.org/encodingFormat'), rdflib.term.URIRef('https://schema.org/contentUrl'), rdflib.term.URIRef('http://www.w3.org/1999/02/22-rdf-syntax-ns#type')}.
  -  [dataset(Titanic) > distribution(Titanic (parquet))] At least one of these properties should be defined: [rdflib.term.URIRef('https://schema.org/sha256'), rdflib.term.URIRef('https://schema.org/md5')]. Existing properties: {rdflib.term.URIRef('https://schema.org/name'), rdflib.term.URIRef('https://schema.org/encodingFormat'), rdflib.term.URIRef('https://schema.org/contentUrl'), rdflib.term.URIRef('http://www.w3.org/1999/02/22-rdf-syntax-ns#type')}.
```

benjelloun commented on datasets/titanic/metadata.json:
```diff
@@ -32,11 +45,11 @@
                 "name": "url",
                 "@type": "ml:Field",
-                "dataType": ["sc:Url", "wd:48277"]
+                "dataType": ["sc:url", "wd:48277"]
```
sc:url is a property, not a type. I think you want sc:URL here.

We discussed with @pierrot0  whether we want to allow properties as well, but I don't think we reached a definitive conclusion on that yet.

marcenacp replied:
What should the verifier check for as far as dataTypes are concerned? How can I verify that a URI defines a dataType?

benjelloun replied:
I think for schema.org and other base types we want to support , it's reasonable to check that the type is one that we know about. It could generate a warning in case of an unknown type.

benjelloun commented on python/format/pyproject.toml:
```diff
@@ -0,0 +1,19 @@
+[project]
+name = "format"
```
Should this be called sth like validator instead?

marcenacp replied:
Seen together: for the moment we keep the current name, because data verification and data generation are very entangled, and it's not clear at this stage how we will split them in two libraries.

benjelloun commented on python/format/src/datasets.py:
```diff
@@ -0,0 +1,55 @@
+        with file.open() as file:
+            file = json.load(file)
+    if not isinstance(file, dict):
+        raise ValueError("The file is not a valid JSON-LD, because it's not a dict.")
```
Nit: Being a "dict" sounds like a Python specific attribute. In JSON terminology, should that be an "object" instead?

benjelloun commented on python/format/src/errors.py:
```diff
@@ -0,0 +1,62 @@
+    """
+    Issues during the validation of the format.
+
+    Issues can either errors (blocking) or warnings (informative).
```
Nit: either *be* errors...

benjelloun commented on python/format/src/errors.py:
```diff
@@ -0,0 +1,62 @@
+        ]:
+            num_issues = len(issues)
+            if num_issues:
+                message += f"\nFound the following {len(issues)} {issue_type} during the validation:"
```
Curious why the \n is at the beginning, not at the end of the messages.

marcenacp replied:
Done.

benjelloun reviewed (changes requested):
Looks good! 
One general question: Do we want so many test files, given the syntax may still change? Is there a way to create fewer of them, and still test all the cases we want to cover?

marcenacp commented:
@benjelloun, thank you for the review!

I also addressed your question on reducing the number of test files. I think we have to keep a minimum of "end-to-end tests" with real data, but I added more unit tests to support other cases.

marcenacp commented:
@pierrot0 Can you please review? Notably, I would like your input on the project's structure:

```
.
|-- python
|   |-- format
|   |   |-- __init__.py
|   |   |-- pyproject.toml
|   |   |-- README.md
|   |   |-- scripts
|   |   |   `-- validate.py
|   |   `-- src
|   |       |-- constants.py
|   |       |-- datasets.py
|   |       |-- datasets_test.py
|   |       |-- errors.py
|   |       |-- errors_test.py
|   |       |-- graphs.py
|   |       |-- graphs_test.py
|   |       |-- __init__.py
|   |       `-- tests
|   |           `-- graphs
|   |               ... *.json files for tests
|   `-- openmlconverter
|       ...
`-- README.md
```

I think:
- We should drop the `src` folder.
- We should have one `datasets_format` library instead of sub-folders (`python/format`).

pierrot0 commented on datasets/titanic/metadata.json:
```diff
@@ -1,11 +1,23 @@
     },
-
-    "@type": "sc:Dataset",
+    "@type": "Dataset",
```
Is there a reason to drop the "sc:" prefix?
This was done to make it more obvious to spot what is coming from where (here schema.org).
If we can keep the prefix, I think it makes the config more readable.

marcenacp replied:
Done.

pierrot0 commented on datasets/titanic/metadata.json:
```diff
@@ -31,10 +43,21 @@
+            },
+            {
+              "name": "label",
+              "description": "Male or female",
```
Female (uppercase F)

marcenacp replied:
Done.

pierrot0 commented on datasets/titanic/metadata.json:
```diff
@@ -31,10 +43,21 @@
-            { "name": "label", "@type": "ml:Field", "dataType": "sc:String" },
+            {
+              "name": "key",
+              "description": "0 or 1",
```
Did you add the description because it was raising an error or a warning?

marcenacp replied:
Not specifying a description currently raises a warning.

pierrot0 commented on datasets/titanic/metadata.json:
```diff
@@ -50,10 +73,21 @@
-            { "name": "label", "@type": "ml:Field", "dataType": "sc:Text" },
+            {
+              "name": "key",
+              "description": "Key in the dataset",
```
To be coherent with the description above, should it be "C, Q, S or ?"?
I am wondering whether we want to recommend adding a description to this "mapping" RecordSets, which are here to map enumeration values to a semantic meaning and human label.

marcenacp replied:
Done.

pierrot0 commented on python/format/README.md:
````diff
@@ -0,0 +1,22 @@
+```bash
+python scripts/validate.py --file ../../datasets/titanic/metadata.json
+```
+
````
Maybe add a sentence on what the behaviour is when the config is valid or invalid: return code is 0 if no errors, otherwise return code is 1 and the first error is displayed, or something alike.

marcenacp replied:
Done.

pierrot0 commented on python/format/src/datasets_test.py:
```diff
@@ -0,0 +1,52 @@
+        # Metadata.
+        [
+            "metadata_missing_property_name.json",
+            'Property "https://schema.org/name" is mandatory, but does not exist.',
```
I think the test should check the context too, or there is a chance that one uses mlfield_missing_property_name.json as an input and get a passing test.

pierrot0 commented on python/format/src/datasets_test.py:
```diff
@@ -0,0 +1,52 @@
+
+
+# End-to-end tests on real data. The data is in `tests/end-to-end/*.json`.
+@pytest.mark.parametrize(
```
Another way to write this test might be to write the output of the validator against the examples to "golden files". The test would then just check whether produced output matches expected output for all examples. 

Pro:
 - context is the expected one.
 - this provides nice examples of what the validator produces, which can then be embedded in the doc.
 - Easier for reviewers to check changes.

Con:
 - we would need a script to update the golden files, and run such a script every time the output format changes. But I think that's doable (I have done something similar in a previous project).

pierrot0 replied:
Also, it's trivial to add examples / tests.

marcenacp replied:
My experience in web development with "snapshot files" is that they are very noisy to update. Particularly, at a time where the format API is likely to change very often, I thought the current solution would be a trade-off between noise/validation.

WDYT of starting with the current solution and update unit tests later if we feel the need?

pierrot0 replied:
SGTM

pierrot0 commented on python/format/src/graphs.py:
```diff
@@ -0,0 +1,243 @@
+    issues: errors.Issues,
+    edges: Edges,
+    dataset_name: str,
+):
```
Can you add a docstring to those functions please?

marcenacp replied:
Done.

pierrot0 commented:
> I think:
> 
> * We should drop the `src` folder.

What would be the gain of dropping the `src` folder?
It feels to me that it might be beneficiary to have a similar structure for all sub-projects (openmlconverter, ...)?

> * We should have one `datasets_format` library instead of sub-folders (`python/format`).

Do you mean that we may want to have multiple github repos?
Or something else?

pierrot0 reviewed (approved):
LGTM
